src/v/kafka/server/audit_log_manager.cc

```diff
@@ -0,0 +1,569 @@
+        }
+        const auto msg = topic.error_message.has_value() ? *topic.error_message
+                                                         : "<no_err_msg>";
+        throw internal_topic_creation_failure(
+          fmt::format("{} - error_code: {}", msg, topic.error_code));
+    }
+}
+
+ss::future<> audit_client::shutdown() {
+    _as.request_abort();
+    _send_sem.broken();
+    co_await _gate.close();
+    co_await _client.stop();
+}
+
+ss::future<>
+audit_client::produce(std::vector<client::record_essence> records) {
+    /// TODO: Produce with acks=1, atm -1 is hardcoded into client
+    const auto records_size = [](const auto& records) {
+        std::size_t size = 0;
+        for (const auto& r : records) {
+            size += r.size_bytes();
+        }
+        return size;
+    };
+
+    try {
+        const auto size_bytes = records_size(records);
+        auto units = co_await ss::get_units(_send_sem, size_bytes);
+        ssx::spawn_with_gate(
+          _gate,
+          [this,
+           units = std::move(units),
+           records = std::move(records)]() mutable {
+              return _client
+                .produce_records(
+                  model::kafka_audit_logging_topic, std::move(records))
+                .discard_result()
+                .handle_exception_type([](const client::partition_error& ex) {
+                    /// TODO: Possible optimization to retry with different
+                    /// partition strategy.
+                    ///
+                    /// If reached here non-mitigatable error occured, or
+                    /// attempts on mitigation had been used up.
+                    vlog(klog.error, "Audit records dropped, reason: {}", ex);
+                })
+                .finally([units = std::move(units)] {});
+          });
+    } catch (const ss::broken_semaphore&) {
+        vlog(
+          klog.debug,
+          "Shutting down the auditor kafka::client, semaphore broken");
+    } catch (const ss::gate_closed_exception&) {
+        vlog(
+          klog.debug, "Shutting down the auditor kafka::client, gate_closed");
+    }
+    co_return;
+}
+
+/// audit_sink
+
+audit_sink::audit_sink(
+  audit_log_manager* audit_mgr,
+  cluster::controller* controller,
+  client::configuration& config) noexcept
+  : _audit_mgr(audit_mgr)
+  , _controller(controller)
+  , _config(config) {}
+
+ss::future<> audit_sink::start() {
+    toggle(true);
+    return ss::now();
+}
+
+ss::future<> audit_sink::stop() {
+    auto gate_closed = _gate.close();
+    _mutex.broken();
+    if (_client) {
+        co_await _client->shutdown();
+        _client.reset(nullptr);
+    }
+    co_await std::move(gate_closed);
+}
+
+ss::future<> audit_sink::produce(std::vector<client::record_essence> records) {
+    auto hold = _gate.hold();
+    auto lock = co_await _mutex.get_units();
+    vassert(_client, "produce() called on a null client");
+    co_await _client->produce(std::move(records));
```

Comment:
what is the mutex protecting in produce? nothing is jumping out at me

Reply:
removed